File: Graphs/graph-dfs.js
// DFS - Graph traversal
// Time: O(V + E)
// Space: O(V)

function dfs(graph, start, visited = new Set()) {
    visited.add(start);
    console.log(start); // or process node
    for (const neighbor of graph[start] || []) {
        if (!visited.has(neighbor)) {
            dfs(graph, neighbor, visited);
        }
    }
}

File: Graphs/graph-dfs.py
# DFS - Graph traversal
# Time: O(V + E)
# Space: O(V)

def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start)  # or process node
    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs(graph, neighbor, visited)
